File: app/src/bls12381.py
# Copyright (C) 2025 Logical Mechanism LLC
# SPDX-License-Identifier: GPL-3.0-only
import secrets
from eth_typing import BLSPubkey, BLSSignature
from src.hashing import generate
from src.constants import H0, F12_DOMAIN_TAG
from py_ecc.bls.g2_primitives import (
    G1_to_pubkey,
    G2_to_signature,
    pubkey_to_G1,
    signature_to_G2,
)
from py_ecc.fields import optimized_bls12_381_FQ as FQ
from py_ecc.fields import optimized_bls12_381_FQ12 as FQ12
from py_ecc.optimized_bls12_381 import (
    G1,
    G2,
    Z1,
    Z2,
    add,
    curve_order,
    multiply,
    neg,
    pairing,
)


def rng() -> int:
    """
    Generates a random hex string of the specified length using the secrets module.

    Returns:
        int: A random number below the field order.
    """
    # Generate a random byte string of the specified length
    return secrets.randbelow(curve_order - 1) + 1


def g1_point(scalar: int) -> str:
    """
    Generates a BLS12-381 point from the G1 generator using scalar multiplication
    and returns it in compressed format.

    Args:
        scalar (int): The scalar value for multiplication.

    Returns:
        bytes: The resulting BLS12-381 G1 point in compressed format.
    """
    return G1_to_pubkey(multiply(G1, scalar)).hex()


def g2_point(scalar: int) -> str:
    """
    Generates a BLS12-381 point from the G2 generator using scalar multiplication
    and returns it in compressed format.

    Args:
        scalar (int): The scalar value for multiplication.

    Returns:
        bytes: The resulting BLS12-381 G2 point in compressed format.
    """
    return G2_to_signature(multiply(G2, scalar)).hex()


def uncompress(element: str) -> tuple:
    """
    Uncompresses a hexadecimal string to a BLS12-381 point.

    Args:
        element (str): The compressed point as a hexadecimal string.

    Returns:
        tuple: The uncompressed point.
    """
    if len(element) == 96:
        return pubkey_to_G1(BLSPubkey(bytes.fromhex(element)))
    else:
        return signature_to_G2(BLSSignature(bytes.fromhex(element)))


def compress(element: tuple) -> str:
    """
    Compresses a BLS12-381 point to a hexadecimal string.

    Args:
        element (tuple): The point to be compressed.

    Returns:
        str: The compressed point as a hexadecimal string.
    """
    if isinstance(element[2], FQ):
        return G1_to_pubkey(element).hex()
    else:
        return G2_to_signature(element).hex()


def scale(element: str, scalar: int) -> str:
    """
    Scales a BLS12-381 point by a given scalar using scalar multiplication.

    Args:
        element (str): The compressed point to be scaled.
        scalar (int): The scalar value for multiplication.

    Returns:
        str: The resulting scaled point.
    """
    return compress(multiply(uncompress(element), scalar))


def invert(element: str) -> str:
    """
    Calculates the inverse of a BLS12-381 point.

    Args:
        element (str): A compressed point.

    Returns:
        str: The resulting combined point.
    """
    return compress(neg(uncompress(element)))


def combine(left_element: str, right_element: str) -> str:
    """
    Combines two BLS12-381 points using addition.

    Args:
        left_element (str): A compressed point.
        right_element (str): A compressed point.

    Returns:
        str: The resulting combined point.
    """
    return compress(add(uncompress(left_element), uncompress(right_element)))


def pair(g1_element: str, g2_element: str, final_exponentiate: bool = True) -> FQ12:
    """
    Compute the pairing operation on elliptic curve points represented as strings.

    Args:
        g2_element (str): A string representation of a point on G2 elliptic curve.
        g1_element (str): A string representation of a point on G1 elliptic curve.
        final_exponentiate (bool, optional): Whether to perform final exponentiation in the pairing computation. Defaults to True.

    Returns:
        FQ12: Result of the pairing operation as an element of the FQ12 field.
    """
    return pairing(uncompress(g2_element), uncompress(g1_element), final_exponentiate)


def fq12_encoding(value: FQ12, domain_tag: str) -> str:
    """
    Encode an FQ12 element into a hex string.

    The element is serialized as the concatenation of its coefficients,
    each written as a 48-byte big-endian integer, and returned as a
    lowercase hex string.

    Args:
        value: The FQ12 field element to encode.

    Returns:
        A hex string representing the byte encoding of the FQ12 element.
    """
    byte_data = b""
    for fq_component in value.coeffs:
        byte_data += int(fq_component).to_bytes(48, "big")
    return generate(byte_data.hex() + domain_tag.encode("utf-8").hex())


def random_fq12(a: int) -> str:
    """
    Derive a deterministic Fq12 element from a scalar and encode it as hex.

    This computes a pairing value:

        kappa = e([a]G1, H0)

    and then serializes/encodes that Fq12 element using `fq12_encoding` with
    the `F12_DOMAIN_TAG`.

    Despite the name, this is not random unless `a` is sampled randomly.

    Args:
        a: Scalar used to multiply the canonical G1 generator before pairing.

    Returns:
        A hex string representing the domain-tagged encoding of the resulting
        Fq12 element.
    """
    kappa = pair(scale(g1_point(1), a), H0)
    return fq12_encoding(kappa, F12_DOMAIN_TAG)


def to_int(hash_digest: str) -> int:
    """
    Interpret a hex digest as a scalar reduced modulo the curve order.

    This is typically used to map a hash output (produced as a hex string)
    into the scalar field used by the protocol:

        c = int(hash_digest, 16) mod curve_order

    Args:
        hash_digest: Hex-encoded digest string (no '0x' prefix expected).

    Returns:
        An integer scalar in the range [0, curve_order - 1].
    """
    return int(hash_digest, 16) % curve_order


def from_int(integer: int) -> str:
    """
    Encode a non-negative integer as a minimal-length big-endian hex string.

    - The encoding is *minimal* (no leading zero bytes).
    - The special case `0` is encoded as `"00"` to ensure a non-empty byte
      representation.

    This is useful for serializing scalars/integers into a compact hex form.

    Args:
        integer: Non-negative integer to encode.

    Returns:
        A hex string representing the integer in big-endian byte order.

    Notes:
        If callers require a fixed width (e.g., always 32 bytes / 64 hex chars),
        this function is intentionally not doing that.
    """
    if integer == 0:
        return "00"
    length = (integer.bit_length() + 7) // 8
    return integer.to_bytes(length, "big").hex()


# identity elements
g1_identity = compress(Z1)
g2_identity = compress(Z2)
gt_identity = FQ12.one()

# curve order
curve_order = curve_order

File: app/gui/fe/src/services/crypto/bls12381.ts
/**
 * BLS12-381 Elliptic Curve Operations
 *
 * Ported from Python: src/bls12381.py
 * Uses @noble/curves for cryptographic operations
 */

import { bls12_381 as bls } from '@noble/curves/bls12-381.js';

// Re-export curve order for use in other modules
export const CURVE_ORDER = bls.fields.Fr.ORDER;

/**
 * Generate a cryptographically secure random scalar below the curve order.
 * Equivalent to Python's rng() function.
 */
export function rng(): bigint {
  // Generate random bytes and reduce modulo curve order
  const randomBytes = new Uint8Array(48);
  crypto.getRandomValues(randomBytes);
  const randomBigInt = bytesToBigInt(randomBytes);
  // Ensure non-zero: (randomBigInt mod (order-1)) + 1
  return (randomBigInt % (CURVE_ORDER - 1n)) + 1n;
}

/**
 * Generate a compressed G1 point from scalar multiplication with the generator.
 * Returns 96 hex characters (48 bytes compressed).
 * Equivalent to Python's g1_point(scalar).
 */
export function g1Point(scalar: bigint): string {
  const point = bls.G1.Point.BASE.multiply(scalar);
  return point.toHex(true); // true = compressed
}

/**
 * Generate a compressed G2 point from scalar multiplication with the generator.
 * Returns 192 hex characters (96 bytes compressed).
 * Equivalent to Python's g2_point(scalar).
 */
export function g2Point(scalar: bigint): string {
  const point = bls.G2.Point.BASE.multiply(scalar);
  return point.toHex(true); // true = compressed
}

/**
 * Uncompress a G1 point from hex string.
 * Input: 96 hex chars (48 bytes compressed)
 */
export function uncompressG1(hex: string): typeof bls.G1.Point.BASE {
  return bls.G1.Point.fromHex(hex);
}

/**
 * Uncompress a G2 point from hex string.
 * Input: 192 hex chars (96 bytes compressed)
 */
export function uncompressG2(hex: string): typeof bls.G2.Point.BASE {
  return bls.G2.Point.fromHex(hex);
}

/**
 * Compress a G1 point to hex string.
 */
export function compressG1(point: typeof bls.G1.Point.BASE): string {
  return point.toHex(true);
}

/**
 * Compress a G2 point to hex string.
 */
export function compressG2(point: typeof bls.G2.Point.BASE): string {
  return point.toHex(true);
}

/**
 * Scale a G1 point by a scalar.
 * Equivalent to Python's scale() function for G1 points.
 */
export function scaleG1(elementHex: string, scalar: bigint): string {
  const point = uncompressG1(elementHex);
  const scaled = point.multiply(scalar);
  return compressG1(scaled);
}

/**
 * Scale a G2 point by a scalar.
 */
export function scaleG2(elementHex: string, scalar: bigint): string {
  const point = uncompressG2(elementHex);
  const scaled = point.multiply(scalar);
  return compressG2(scaled);
}

/**
 * Scale a point (auto-detect G1 or G2 based on hex length).
 * G1: 96 hex chars, G2: 192 hex chars
 */
export function scale(elementHex: string, scalar: bigint): string {
  if (elementHex.length === 96) {
    return scaleG1(elementHex, scalar);
  } else if (elementHex.length === 192) {
    return scaleG2(elementHex, scalar);
  } else {
    throw new Error(`Invalid element length: ${elementHex.length}. Expected 96 (G1) or 192 (G2).`);
  }
}

/**
 * Negate a G1 point.
 * Equivalent to Python's invert() function.
 */
export function invertG1(elementHex: string): string {
  const point = uncompressG1(elementHex);
  const negated = point.negate();
  return compressG1(negated);
}

/**
 * Negate a G2 point.
 */
export function invertG2(elementHex: string): string {
  const point = uncompressG2(elementHex);
  const negated = point.negate();
  return compressG2(negated);
}

/**
 * Combine two G1 points using addition.
 * Equivalent to Python's combine() function.
 */
export function combineG1(leftHex: string, rightHex: string): string {
  const left = uncompressG1(leftHex);
  const right = uncompressG1(rightHex);
  const combined = left.add(right);
  return compressG1(combined);
}

/**
 * Combine two G2 points using addition.
 */
export function combineG2(leftHex: string, rightHex: string): string {
  const left = uncompressG2(leftHex);
  const right = uncompressG2(rightHex);
  const combined = left.add(right);
  return compressG2(combined);
}

/**
 * Combine two points (auto-detect G1 or G2).
 */
export function combine(leftHex: string, rightHex: string): string {
  if (leftHex.length !== rightHex.length) {
    throw new Error('Cannot combine points of different types');
  }
  if (leftHex.length === 96) {
    return combineG1(leftHex, rightHex);
  } else if (leftHex.length === 192) {
    return combineG2(leftHex, rightHex);
  } else {
    throw new Error(`Invalid element length: ${leftHex.length}`);
  }
}

/**
 * Interpret a hex digest as a scalar reduced modulo the curve order.
 * Equivalent to Python's to_int() function.
 */
export function toInt(hashDigest: string): bigint {
  const bigInt = BigInt('0x' + hashDigest);
  return bigInt % CURVE_ORDER;
}

/**
 * Encode a non-negative integer as a minimal-length big-endian hex string.
 * Equivalent to Python's from_int() function.
 */
export function fromInt(integer: bigint): string {
  if (integer === 0n) {
    return '00';
  }
  let hex = integer.toString(16);
  // Ensure even length
  if (hex.length % 2 !== 0) {
    hex = '0' + hex;
  }
  return hex;
}

/**
 * Convert bytes (Uint8Array) to BigInt
 */
function bytesToBigInt(bytes: Uint8Array): bigint {
  let hex = '';
  for (const byte of bytes) {
    hex += byte.toString(16).padStart(2, '0');
  }
  return BigInt('0x' + hex);
}

/**
 * Convert hex string to Uint8Array
 */
export function hexToBytes(hex: string): Uint8Array {
  if (hex.length % 2 !== 0) {
    hex = '0' + hex;
  }
  const bytes = new Uint8Array(hex.length / 2);
  for (let i = 0; i < bytes.length; i++) {
    bytes[i] = parseInt(hex.substr(i * 2, 2), 16);
  }
  return bytes;
}

/**
 * Convert Uint8Array to hex string
 */
export function bytesToHex(bytes: Uint8Array): string {
  return Array.from(bytes)
    .map((b) => b.toString(16).padStart(2, '0'))
    .join('');
}

/**
 * G1 identity element (point at infinity)
 */
export const G1_IDENTITY = compressG1(bls.G1.Point.ZERO);

/**
 * G2 identity element (point at infinity)
 */
export const G2_IDENTITY = compressG2(bls.G2.Point.ZERO);

/**
 * G1 generator point
 */
export const G1_GENERATOR = g1Point(1n);

/**
 * G2 generator point
 */
export const G2_GENERATOR = g2Point(1n);
